From 91df8f3c99e60ab052994e5b01b1cecb1bc522fa Mon Sep 17 00:00:00 2001
From: Ian Macdonald <ian@caliban.org>
Date: Sat, 24 Nov 2018 19:32:09 +0100
Subject: [PATCH 09/17] fs/sdfat: fix build when SDFAT_DEBUG unset

---
 fs/sdfat/blkdev.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/fs/sdfat/blkdev.c b/fs/sdfat/blkdev.c
index 264c670df0f0..d0e0af835005 100644
--- a/fs/sdfat/blkdev.c
+++ b/fs/sdfat/blkdev.c
@@ -96,7 +96,9 @@ s32 bdev_check_bdi_valid(struct super_block *sb)
 			fsi->prev_eio |= SDFAT_EIO_BDI;
 			sdfat_log_msg(sb, KERN_ERR, "%s: block device is "
 				"eliminated.(bdi:%p)", __func__, sb->s_bdi);
+#ifdef CONFIG_SDFAT_DEBUG
 			sdfat_debug_warn_on(1);
+#endif
 		}
 		return -ENXIO;
 	}
@@ -159,7 +161,9 @@ s32 bdev_mread(struct super_block *sb, u64 secno, struct buffer_head **bh, u64 n
 	if (!(fsi->prev_eio & SDFAT_EIO_READ)) {
 		fsi->prev_eio |= SDFAT_EIO_READ;
 		sdfat_log_msg(sb, KERN_ERR, "%s: No bh. I/O error.", __func__);
+#ifdef CONFIG_SDFAT_DEBUG
 		sdfat_debug_warn_on(1);
+#endif
 	}
 
 	return -EIO;
@@ -213,7 +217,9 @@ s32 bdev_mwrite(struct super_block *sb, u64 secno, struct buffer_head *bh, u64 n
 	if (!(fsi->prev_eio & SDFAT_EIO_WRITE)) {
 		fsi->prev_eio |= SDFAT_EIO_WRITE;
 		sdfat_log_msg(sb, KERN_ERR, "%s: No bh. I/O error.", __func__);
+#ifdef CONFIG_SDFAT_DEBUG
 		sdfat_debug_warn_on(1);
+#endif
 	}
 
 	return -EIO;
-- 
2.17.1

